sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -0,0 +1,231 @@
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task StopAsync()
+        {
+            if (RunningTask != null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask != null)
+                    {
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        await RunningTask.ConfigureAwait(false);
+                        RunningTask = null;
+
+                        await Task.WhenAll(PartitionPumps.Select(kvp => kvp.Value.StopAsync())).ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   The main loop of an event processor.  It loops through every owned <see cref="PartitionPump" />, checking
+        ///   its status and creating a new one if necessary.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        /// <remarks>
+        ///   The actual goal of this method is to perform load balancing between multiple <see cref="EventProcessor" />
+        ///   instances, but this feature is currently out of the scope of the current preview.
+        /// </remarks>
+        ///
+        private async Task RunAsync(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                await Task.WhenAll(PartitionPumps
+                    .Where(kvp => !kvp.Value.IsRunning)
+                    .Select(kvp =>
+                    {
+                        var partitionId = kvp.Key;
+
+                        var partitionContext = new PartitionContext(InnerClient.EventHubPath, ConsumerGroup, partitionId);
+                        var checkpointManager = new CheckpointManager(partitionContext, PartitionManager, Identifier);
+
+                        var partitionProcessor = PartitionProcessorFactory.CreatePartitionProcessor(partitionContext, checkpointManager);
+
+                        var partitionPump = new PartitionPump(InnerClient, ConsumerGroup, partitionId, partitionProcessor, Options);
+                        PartitionPumps.TryUpdate(partitionId, partitionPump, partitionPump);
+
+                        return partitionPump.StartAsync();
+                    })).ConfigureAwait(false);
+
+                // Wait 1 second before the next verification.
+
+                await Task.Delay(1000).ConfigureAwait(false);
```

Comment:
nit:  You probably want to pass the cancellation token into the `WhenAll` and the `Delay`.     Be aware that if you do, they'll potentially throw a `TaskCanceledExcetption` that you'll need to handle (or choose to ignore and let bubble.)